### ISSUE TITLE
Usage of Spring Data Redis 1.7.1 which has critical bug https://jira.spring.io/browse/DATAREDIS-502

### ISSUE THREAD

**Issue by gennadiy-gerasimov** [spring-projects/spring-session#756]
Hi, 

Spring Session uses Spring Data Redis 1.7.1 which has critical bug already fixed in 1.7.3 https://jira.spring.io/browse/DATAREDIS-502

I haven't seen any related pull requests for 1.4.0 M1. Are there any plans to update Spring Data Redis?

P.S. I do not know gradle well but i suppose that "springDataReleaseTrainVersion=Kay-BUILD-SNAPSHOT" means that current master branch points to the latest Spring Data Redis version  2.0 M3 (Kay). If so is it possible to have Snapshot of Spring Session which uses  2.0 M3 (Kay)?

**Comment by jxblum** [comment]
For _Spring Session_ **1.4.x** (next is `1.4.0.M1`), _Spring Data Redis_ has already been updated to [1.8.0.RC1](https://github.com/spring-projects/spring-session/blob/1.4.x/gradle.properties#L25).  This should probably be updated to the latest version of SD Redis on the 1.8.x line (currently, `1.8.3.RELEASE`).

For _Spring Session_ **master** (next is `2.0.0.M1`), _Spring Data Redis_ has been updated to _Spring Data_ **Kay**, which includes the latest, _Spring Data Redis_ 2.0.0.  _Spring Session_ 2.0 has been [rebased on the SD _Release Train_](https://github.com/spring-projects/spring-session/blob/master/gradle.properties#L21) (going forward) to keep all SD modules from the SD _Release Train_ (namely, SD GemFire, SD MongoDB and SD Redis) in sync and up-to-date.  Repositories supported by _Spring Session_ that are not part of the SD _Release Train_ (as they are public modules, e.g. Hazelcast) is not subject to the same in-sync updates, so must be maintained separately.  The other reason _Spring Session_ (as of 2.0) has been sync'd with the SD _Release Train_ is because _Spring Session_ is also pulled in by _Spring Boot_, which also has a dependency on the SD _Release Train_, and, as of Spring Session 2, which is based on Spring Boot 2, both are based on SD **Kay**.

As for _Spring Session_ 1.3.x, _Spring Session_ will stay fixed on SD Redis **1.7.x**, but will be updated to the latest maintenance releases.  For SD Redis, that is currently `1.7.10.RELEASE`.  And, as you can see from [DATAREDIS-502](https://jira.spring.io/browse/DATAREDIS-502), this issue was fixed in SD Redis `1.7.3.RELEASE (Hopper)`.

I will make the necessary updates.

Cheers,
John

**Comment by jxblum** [comment]
Damn it!  Now, I remember why _Spring Session_ `1.3.x` could not be upgraded beyond _Spring Data Redis_ `1.7.1.RELEASE`.  Test failures (a method used by SDR was removed in core _Spring_)...

```
org.springframework.session.data.redis.config.annotation.web.http.RedisHttpSessionConfigurationTests > resolveValue FAILED
    org.springframework.beans.factory.UnsatisfiedDependencyException at RedisHttpSessionConfigurationTests.java:72
        Caused by: org.springframework.beans.factory.BeanCreationException at RedisHttpSessionConfigurationTests.java:72
            Caused by: java.lang.NoSuchMethodError at RedisHttpSessionConfigurationTests.java:72
```

This had to do with a compatibility issue between the core _Spring Framework_ 
version [set](https://github.com/spring-projects/spring-session/blob/1.3.0.RELEASE/gradle.properties#L19) by _Spring Session_ 1.3 (i.e. `4.3.4.RELEASE`) and the core _Spring Framework_ version expected by _Spring Data Redis_ 1.7 (indirectly [determined by _Spring Data Build_](https://github.com/spring-projects/spring-data-redis/blob/1.7.1.RELEASE/pom.xml#L14-L15), and then [this](https://github.com/spring-projects/spring-data-build/blob/1.8.1.RELEASE/parent/pom.xml#L106)).

While the _Spring Framework_ versions are not that far removed (4.2.5 -> 4.3.4).  I think a core _Spring_ API method available in 4.2 used by _Spring Data Redis_ (1.7.x) was removed in core _Spring_ 4.3.

I should have remembered this since I tried to upgrade SD Redis previously on `1.3.x`.  I will try to dig up the earlier comments made regarding this.

**Comment by jxblum** [comment]
Found it!  The details are explained in the first half of this [comment](https://github.com/spring-projects/spring-session/issues/677#issuecomment-271731284).

@rwinch will need to chime in here as to how he wants to handle this.

There are a few options...

1. (Least likely) Drop support for _Spring_ 3 in _Spring Session_ `1.3.x` (starting from the `1.3.1.RELEASE`).  Core _Spring Framework_ `3.2.x` is effectively in its EOL anyway, at least there are no planned maintenance versions beyond `3.2.18.RELEASE` that I am aware of.

2. Wait for _Spring Session_ `1.4.0.M1`.  I have just [upgraded](https://github.com/spring-projects/spring-session/blob/1.4.x/gradle.properties#L25) _Spring Session_ `1.4` to _Spring Data Redis_ `1.8.3.RELEASE`, the latest version in _Spring Data_ **Ingalls**.

3. Finally, you can add an exclusion to the version (i.e. `1.7.1.RELEASE`) of _Spring Data Redis_ pulled in by _Spring Session_ `1.3.0.RELEASE` and add an explicit dependency on a later version in your application Maven POM file or Gradle build file.  For example...

```
compile ("org.springframework.session:spring-session:1.3.0.RELEASE") {
    exclude group: "org.springframework.data", module: 'spring-data-redis'
}

compile "org.springframework.data:spring-data-redis:1.7.10.RELEASE"
```

You may be importing the `spring-session-data-redis` module, specifically.  If so, you will need to switch to the above.  I don't foresee any problems running the latest SD _Redis_ 1.7 with _Spring Session_ 1.3.

**Comment by jxblum** [comment]
As further confirmation, when I [set](https://github.com/spring-projects/spring-session/blob/1.3.x/gradle.properties#L26) the _Spring Data Redis_ version to "`1.7.10.RELEASE`" and run the _Spring Session_ `1.3.x` build...

```
$ gradlew clean build install -x spring3Test
```

NOTE: `-x spring3Test` effectively skips the _Spring_ **3** compatibility testing

Then the _Spring Session_ Test Suite **passes**!

```
...
:spring-session:install
:spring-session-data-gemfire:install
:spring-session-data-mongo:install
:spring-session-data-redis:install
:spring-session-hazelcast:install
:spring-session-jdbc:install

BUILD SUCCESSFUL

Total time: 8 mins 28.248 secs
```

Therefore, functionally, _Spring Data Redis_ `1.7.10.RELEASE` should work just fine with _Spring Session_ `1.3.0.RELEASE`.

Still, I **cannot** check in this change for _Spring Session_ **1.3.x** (for `1.3.1.RELEASE`).

@rwinch, _Spring Session_ lead, must decide how to proceed.

Thanks!

**Comment by jxblum** [comment]
Final note, both _Spring Session_ `master` and `1.4.x` are **up-to-date** as of this comment.

**Comment by gennadiy-gerasimov** [comment]
The option #3 works for me. Thanks for help!

**Comment by rwinch** [comment]
This is now fixed in 1.3.x. To get spring3Tests I explicitly override the Spring Data Redis version for those tests. This is reasonable since we are overriding the Spring Framework version.

**NOTE** You can override it simply by providing the spring-data-redis dependency explicitly. There is no need for the exclusion since explicit dependencies are preferred over transitive dependencies.
